### DISCUSSION TITLE
Metadata: Use UTC offset if the actual time zone is unknown

### DISCUSSION THREAD

**Discussion by shagog** [photoprism/photoprism#3780] [Q&A] [answered]
#### 1. What is not working as documented?

*Probably related to https://github.com/photoprism/photoprism/issues/2320 but without GPS specifics.*

Probably `TakenAt` value is not assigned correctly in sidecar, when indexing .heic.

When indexing .heic images, taken on iPhones *with disabled GPS for camera*, the `TakenAt` value in sidecar file is set to `DateTimeOriginal` (or `CreateDate` ?) value from exif **adding Z timezone**, without respect to the  `OffsetTime`, `OffsetTimeOriginal`, `OffsetTimeDigitized`, `SubSecCreateDate` and `SubSecDateTimeOriginal` tags, even when they contain offset value.

----
Consequent issue:

*In https://github.com/photoprism/photoprism/issues/2320#issuecomment-1130471827 it was mentioned, that photos are sorted by local times, not by UTC. So this is considered expected behavior?*

This leads to images, even having proper time offset, being displayed not in 'natural' order.

For example, if there are 3 pictures (pls see visual example below):
Pic1: 2023-10-02 13:20:17 +02:00
Pic2: 2023-10-02 12:22:38 +01:00
Pic3: 2023-10-02 13:25:27 +02:00

The order they are displayed in Photoprism will be: Pic3, Pic1, Pic2 (Newest first sort)
Expected order: Pic3, Pic2, Pic1 (Newest first sort)

For the Oldest first sort it would be: Pic2, Pic1, Pic3
Expected order: Pic1, Pic2, Pic3

The order they were taken in ‘real life’ is: Pic1, Pic2, Pic3

Still, it's probably better to keep time zone in the sidecar file correct.

Cases, when sorting issue matters: 
1. Library contains photos taken by different people/devices in different time zones at about same moment. 
2. One’s taken photos before and after crossing time zone line where time between photos is less, than the time zones difference
3. Library contains photos taken by different people/devices in same location at about same moment, while having phones’ time locked to their ‘home’ time zone and not the local one

#### 2. How can we reproduce it?

0. Take photos sequentially on iPhone in different time zones with GPS disable for camera (attached archive with sample pictures)
1. Upload photos taken to Originals folder. 
2. Index uploaded photos
3. Observe that corresponding sidecar files contain Z as timezone
4. Go to Search. Sort by Newesr or Latest first
5. Observe that photos are sorted as if their ‘local’ time was utc or if they had same time offset

For the sake of the example, photos were taken on same device, however behaviour is consistent if photos were taken on several different iphones.

#### 3. What behaviour do you expect?

I would expect for pictures without GPS coordinates, but with time offset present, to have corresponding offset assigned. So for the example above it would be +02:00 for pic1 and pic3 and +01:00 for pic 2.
In case it is necessary to store `TakenAt` in UTC, probably it would be more correct to adjust the time according to the given offset.

Also, would expect pictures to be sorted by 'utc' time (meaning ‘local’ time +/- offset), so they will be ordered as the they were taken in real life.

#### 4. What could be the cause of your problem?

Allegedly Apple populates the exif data not in supposed way (see example below), so time offset is not embedded into `DateTimeOriginal` or `CreateDate`, but provided as additional `OffsetTime`, `OffsetTimeOriginal` and `OffsetTimeDigitized` fields, without providing any time zone info.
 
If I’m looking at the right place (https://github.com/photoprism/photoprism/blob/release/internal/meta/exif.go), those tags are not being read from exif (also they are not mentioned in https://www.photoprism.app/kb/metadata).

However, offset is embedded in `SubSecCreateDate` and `SubSecDateTimeOriginal`, which should prevail over `DateTimeOriginal`, but as far as I understood, the offset is not being read from those fields by PhotoPrism (? Not really sure about that).

Reduced exif for Pic1 (`exiftool -time:all -j`):

```
[{
  "SourceFile": "IMG_9395.heic",
  "FileModifyDate": "2023:10:02 16:28:10+02:00",
  "FileAccessDate": "2023:10:02 13:52:51+02:00",
  "FileInodeChangeDate": "2023:10:02 13:37:35+02:00",
  "ModifyDate": "2023:10:02 13:20:17",
  "DateTimeOriginal": "2023:10:02 13:20:17",
  "CreateDate": "2023:10:02 13:20:17",
  "OffsetTime": "+02:00",
  "OffsetTimeOriginal": "+02:00",
  "OffsetTimeDigitized": "+02:00",
  "SubSecTimeOriginal": 568,
  "SubSecTimeDigitized": 568,
  "ProfileDateTime": "2022:01:01 00:00:00",
  "SubSecCreateDate": "2023:10:02 13:20:17.568+02:00",
  "SubSecDateTimeOriginal": "2023:10:02 13:20:17.568+02:00",
  "SubSecModifyDate": "2023:10:02 13:20:17+02:00"
}]
```

<details>

<summary>Full exif:</summary>

```
[{
  "SourceFile": "IMG_9395.heic",
  "ExifToolVersion": 11.88,
  "FileName": "IMG_9395.heic",
  "Directory": ".",
  "FileSize": "2.9 MB",
  "FileModifyDate": "2023:10:02 16:28:10+02:00",
  "FileAccessDate": "2023:10:02 13:44:01+02:00",
  "FileInodeChangeDate": "2023:10:02 13:37:35+02:00",
  "FilePermissions": "rw-rw-r--",
  "FileType": "HEIC",
  "FileTypeExtension": "heic",
  "MIMEType": "image/heic",
  "MajorBrand": "High Efficiency Image Format HEVC still image (.HEIC)",
  "MinorVersion": "0.0.0",
  "CompatibleBrands": ["mif1","MiPr","miaf","MiHB","heic"],
  "HandlerType": "Picture",
  "PrimaryItemReference": 49,
  "ExifByteOrder": "Big-endian (Motorola, MM)",
  "Make": "Apple",
  "Model": "iPhone 13",
  "Orientation": "Rotate 90 CW",
  "XResolution": 72,
  "YResolution": 72,
  "ResolutionUnit": "inches",
  "Software": "16.6.1",
  "ModifyDate": "2023:10:02 13:20:17",
  "HostComputer": "iPhone 13",
  "ExposureTime": "1/163",
  "FNumber": 1.6,
  "ExposureProgram": "Program AE",
  "ISO": 40,
  "ExifVersion": "0232",
  "DateTimeOriginal": "2023:10:02 13:20:17",
  "CreateDate": "2023:10:02 13:20:17",
  "OffsetTime": "+02:00",
  "OffsetTimeOriginal": "+02:00",
  "OffsetTimeDigitized": "+02:00",
  "ShutterSpeedValue": "1/163",
  "ApertureValue": 1.6,
  "BrightnessValue": 5.937361762,
  "ExposureCompensation": 0,
  "MeteringMode": "Multi-segment",
  "Flash": "Off, Did not fire",
  "FocalLength": "5.1 mm",
  "SubjectArea": "2011 1508 2323 1330",
  "RunTimeFlags": "Valid",
  "RunTimeValue": 515774842809958,
  "RunTimeScale": 1000000000,
  "RunTimeEpoch": 0,
  "AccelerationVector": "0.01318094506 0.00392230414 -1.016510248",
  "Warning": "[minor] Bad format (16) for MakerNotes entry 13",
  "SubSecTimeOriginal": 568,
  "SubSecTimeDigitized": 568,
  "ColorSpace": "Uncalibrated",
  "ExifImageWidth": 4032,
  "ExifImageHeight": 3024,
  "SensingMethod": "One-chip color area",
  "SceneType": "Directly photographed",
  "ExposureMode": "Auto",
  "WhiteBalance": "Auto",
  "FocalLengthIn35mmFormat": "26 mm",
  "LensInfo": "1.539999962-5.1mm f/1.6-2.4",
  "LensMake": "Apple",
  "LensModel": "iPhone 13 back dual wide camera 5.1mm f/1.6",
  "CompositeImage": "General Composite Image",
  "ProfileCMMType": "Apple Computer Inc.",
  "ProfileVersion": "4.0.0",
  "ProfileClass": "Display Device Profile",
  "ColorSpaceData": "RGB ",
  "ProfileConnectionSpace": "XYZ ",
  "ProfileDateTime": "2022:01:01 00:00:00",
  "ProfileFileSignature": "acsp",
  "PrimaryPlatform": "Apple Computer Inc.",
  "CMMFlags": "Not Embedded, Independent",
  "DeviceManufacturer": "Apple Computer Inc.",
  "DeviceModel": "",
  "DeviceAttributes": "Reflective, Glossy, Positive, Color",
  "RenderingIntent": "Perceptual",
  "ConnectionSpaceIlluminant": "0.9642 1 0.82491",
  "ProfileCreator": "Apple Computer Inc.",
  "ProfileID": "ecfda38e388547c36db4bd4f7ada182f",
  "ProfileDescription": "Display P3",
  "ProfileCopyright": "Copyright Apple Inc., 2022",
  "MediaWhitePoint": "0.96419 1 0.82489",
  "RedMatrixColumn": "0.51512 0.2412 -0.00105",
  "GreenMatrixColumn": "0.29198 0.69225 0.04189",
  "BlueMatrixColumn": "0.1571 0.06657 0.78407",
  "RedTRC": "(Binary data 32 bytes, use -b option to extract)",
  "ChromaticAdaptation": "1.04788 0.02292 -0.0502 0.02959 0.99048 -0.01706 -0.00923 0.01508 0.75168",
  "BlueTRC": "(Binary data 32 bytes, use -b option to extract)",
  "GreenTRC": "(Binary data 32 bytes, use -b option to extract)",
  "HEVCConfigurationVersion": 1,
  "GeneralProfileSpace": "Conforming",
  "GeneralTierFlag": "Main Tier",
  "GeneralProfileIDC": "Main Still Picture Profile",
  "GenProfileCompatibilityFlags": "Main Still Picture, Main 10, Main",
  "ConstraintIndicatorFlags": "176 0 0 0 0 0",
  "GeneralLevelIDC": "90 (level 3.0)",
  "MinSpatialSegmentationIDC": 0,
  "ParallelismType": 0,
  "ChromaFormat": "4:2:0",
  "BitDepthLuma": 8,
  "BitDepthChroma": 8,
  "AverageFrameRate": 0,
  "ConstantFrameRate": "Unknown",
  "NumTemporalLayers": 1,
  "TemporalIDNested": "No",
  "ImageWidth": 4032,
  "ImageHeight": 3024,
  "ImageSpatialExtent": "4032x3024",
  "Rotation": 270,
  "ImagePixelDepth": "8 8 8",
  "MediaDataSize": 2994790,
  "MediaDataOffset": 3419,
  "RunTimeSincePowerUp": "5 days 23:16:15",
  "Aperture": 1.6,
  "ImageSize": "4032x3024",
  "Megapixels": 12.2,
  "ScaleFactor35efl": 5.1,
  "ShutterSpeed": "1/163",
  "SubSecCreateDate": "2023:10:02 13:20:17.568+02:00",
  "SubSecDateTimeOriginal": "2023:10:02 13:20:17.568+02:00",
  "SubSecModifyDate": "2023:10:02 13:20:17+02:00",
  "CircleOfConfusion": "0.006 mm",
  "FOV": "69.4 deg",
  "FocalLength35efl": "5.1 mm (35 mm equivalent: 26.0 mm)",
  "HyperfocalDistance": "2.76 m",
  "LightValue": 10.0
}]
```

</details>

#### 5. Can you provide us with example files for testing, error logs, or screenshots?

Please see included archive with sample set.

[test_set.zip](https://github.com/photoprism/photoprism/files/12782668/test_set.zip)

When they are being indexed, they are sorted like this:
![GUI_sorted](https://github.com/photoprism/photoprism/assets/146329185/6cdedbed-39d3-4fc5-8829-5a972c751154)

In GUI the date-time displayed is same as `DateTimeOriginal`, with time zone set to ‘Local Time’.

![GUI_datetime](https://github.com/photoprism/photoprism/assets/146329185/0ad0b74c-e823-4f2f-8615-1bed0363840a)

I do understand that time offset != time zone, so this is expected representation in GUI. I would suggest assigning any of the corresponding named time zones for such images or have other [‘lettered’ timezones](https://en.wikipedia.org/wiki/Military_time_zone) in addition to Z. Not sure if this would be a viable solution though. 

The sidecar .yml files created automatically for those images:

<details>

<summary>Pic1</summary>

```
TakenAt: 2023-10-02T13:20:17Z
TakenAtLocal: 2023-10-02T13:20:17Z
TakenSrc: meta
UID: ps1wo7f30wabylni
Type: image
Title: Unknown / 2023
Year: 2023
Month: 10
Day: 2
ISO: 40
Exposure: 1/163
FNumber: 1.6
FocalLength: 26
Quality: 3
Details:
  Keywords: brown
  Software: 16.6.1
  SoftwareSrc: meta
CreatedAt: 2023-10-02T14:28:27Z
UpdatedAt: 2023-10-02T14:28:28.126616854Z
```

</details>

<details>

<summary>Pic2</summary>

```
TakenAt: 2023-10-02T12:22:38Z
TakenAtLocal: 2023-10-02T12:22:38Z
TakenSrc: meta
UID: ps1wo7fxo5agf9ar
Type: image
Title: Unknown / 2023
Year: 2023
Month: 10
Day: 2
ISO: 32
Exposure: 1/122
FNumber: 1.6
FocalLength: 26
Quality: 3
Details:
  Keywords: brown
  Software: 16.6.1
  SoftwareSrc: meta
CreatedAt: 2023-10-02T14:28:27Z
UpdatedAt: 2023-10-02T14:28:28.237348368Z
```

</details>

<details>

<summary>Pic3</summary>

```
TakenAt: 2023-10-02T13:25:27Z
TakenAtLocal: 2023-10-02T13:25:27Z
TakenSrc: meta
UID: ps1wo7f3by6qkg6h
Type: image
Title: Unknown / 2023
Year: 2023
Month: 10
Day: 2
ISO: 40
Exposure: 1/121
FNumber: 1.6
FocalLength: 26
Quality: 3
Details:
  Keywords: green
  Software: 16.6.1
  SoftwareSrc: meta
CreatedAt: 2023-10-02T14:28:27Z
UpdatedAt: 2023-10-02T14:28:28.281956512Z
```

</details>
 

#### 6. Which software versions do you use?

(a) PhotoPrism Architecture & Build Number:
AMD64, Build 230923-e59851350

(b) Database Type & Version: 
MariaDB 10.11

(c) Operating System Types & Versions: 
Photoprism runs in Docker, image with `latest` tag as of 30.09.2023

Host OS - Debian 11

(d) Browser Types & Versions: 
Chrome 114, Safari 16.6 (Mac OS and iOS), but not applicable, I believe

(e) Ad Blockers, Browser Plugins, and/or Firewall Software?
Not applicable, I believe 

(f)
Hardware and OS, producing images:
iPhones with IOS 16.x with GPS disabled for photos.
The time and time zone could be set either automatically or manual, doesn’t affect.

Haven’t tried with IOS 17 yet. 

#### 7. On what kind of device is PhotoPrism installed?

I suppose this should not be hardware-dependent, but since this is required for import/indexing-related issues:

(a) Device / Processor Type: 
AMD Ryzen 5 5600X

(b) Physical Memory & Swap Space in GB:
32GB + 4GB swap

(c) Storage Type: HDD, SSD, RAID, USB, Network Storage:
PhotoPrism runs from SSD and has its files on SSD, the originals folder is mounted from HDD

**Comment by lastzero** [comment, marked as answer]
@shagog Your request has been implemented and will be available with our next update, scheduled for today or tomorrow. We appreciate your help with testing these changes! :)

**Reply by lastzero** [reply]
:point_right: https://docs.photoprism.app/release-notes/

**Reply by shagog** [reply]
Thank you a lot!
I've tried on a small dataset for now and it works. 
(**disclaimer for someone with same issue: you DO need to run rescan first**) 

However, there are several nuances (not sure if they are issues):

1. `TakenAt` is correct, but `TakenAtLocal` still has `Z` as a time zone. So it's like that for pic1: 
```
TakenAt: 2023-10-02T11:20:17Z
TakenAtLocal: 2023-10-02T13:20:17Z
<...>
TimeZone: UTC+2
```
Wouldn't it be correct to drop `Z` at `TakenAtLocal`, as it is not in UTC?

2. (minor) There is small inconsistency in GUI: 
In the card view timezone is represented as GMT+x, while in the "Details" and in sidecar yml it is represented as UTC+x. Is it intended or is it a typo?

![Screenshot_details](https://github.com/photoprism/photoprism/assets/146329185/5c1977a3-c2e7-42a8-97ce-4c1d94cde2f4)


3. (question) Am I right to assume, that Newest/Latest sort is performed based on `TakenAtLocal` and this is expected behavior? Because even with the new correct  `TakenAt` and `TimeZone` values, the sorting is still not in "natural" order, but as if photos were taken in same time zone (pls see below)

![Screenshot_sorting](https://github.com/photoprism/photoprism/assets/146329185/9dbb2ecd-a16a-441e-8ec1-3ce0af5e5447)


I'll run rescan on the bigger dataset and will tell if there'll be anything else, but it takes some time :)

**Reply by lastzero** [reply]
Z is technically not correct for the local time but "required" for serialization using the standard tools and libraries. This is because timestamps are saved in the database without timezone while our software libraries require a timezone and also automatically add it when values are serialized/saved.

Sorting behavior is based on the feedback / requests from our community. Also note that not all pictures have a known UTC time even with these changes. So these pictures would be "wrongly" sorted anyways. If it happens that we somehow have a lot of time and nothing else to do, we can add yet another config option to customize sorting by time.

**Reply by shagog** [reply]
Got it, thanks!

I've run rescan on the bigger dataset, it looks to me that everything works good so far. Videos and live photos were also handled correctly.

I think this discussion could be marked as resolved.

The GMT/UTC inconsistency doesn't bother me at all, so its fully at your discretion.

**Reply by lastzero** [reply]
We could add a special UTC sort option for users with this requirement. It might be hidden by default, like the "random" sort order that's implemented but not documented.

**Reply by shagog** [reply]
Just one question: are hidden options only available for the search results? Tried to use the "random" option as query param `?view=cards&order=random` inside the album view, but the results were still sorted as set in album's settings. However it works fine if done in Search interface with same album specified as filter.

If so, I would suggest a bit different approach: may be it could be an option in compose file (environment variable like Experimental flag, probably not documented one) that would alter the behavior of default Oldest/Newest sort options to use UTC (or "field that is supposed to be UTC")?

I understand that it might be not so common requirement and if this approach is too much trouble - the hidden option like "random" would be great!

**Reply by lastzero** [reply]
You would also need to change the creation and updating of the following indexes in the files table (unless you are okay with less performance and significant changes to the SQL queries):
https://github.com/photoprism/photoprism/blob/c40bf9a32afb47aef87660ab61ee93285bb09855/internal/entity/file.go#L46-L47

**Reply by shagog** [reply]
If I'm reading correctly, both `media_id` and `time_index` are dependending at `photo_taken_at`
So would something like that do the trick or should I dig deeper?


lines https://github.com/photoprism/photoprism/blob/c40bf9a32afb47aef87660ab61ee93285bb09855/internal/entity/file.go#L129-L131
could be changed to 

```Go
	case MySQL:
		switch TimeSortingOption {
			case UTCSort:
				// Use pictures.taken_at field (assuming it is in UTC) if available, defaulting to taken_at_local if NULL
				Log("files", "regenerate photo_taken_at",
					Db().Exec("UPDATE files JOIN ? p ON p.id = files.photo_id SET files.photo_taken_at = CASE WHEN p.taken_at IS NOT NULL THEN p.taken_at ELSE p.taken_at_local END  WHERE ?",
						gorm.Expr(photosTable), updateWhere).Error)
			
			default:
				//Keeping taken_at_local, original request
				Log("files", "regenerate photo_taken_at",
					Db().Exec("UPDATE files JOIN ? p ON p.id = files.photo_id SET files.photo_taken_at = p.taken_at_local WHERE ?",
						gorm.Expr(photosTable), updateWhere).Error)		
		}

```

And for the SQLite 
https://github.com/photoprism/photoprism/blob/c40bf9a32afb47aef87660ab61ee93285bb09855/internal/entity/file.go#L140-L143

could be changed to smth like

```Go
	case SQLite3:
		switch TimeSortingOption {
			case UTCSort:
				// Use pictures.taken_at field (assuming it is in UTC) if available, defaulting to taken_at_local if NULL
				// Request below is not optimal, but not sure how to optimize it in SQLite
				Log("files", "regenerate photo_taken_at",
					Db().Exec("UPDATE files SET photo_taken_at = CASE WHEN (SELECT p.taken_at FROM ? p WHERE p.id = photo_id) as ta IS NOT NULL THEN ta ELSE (SELECT p.taken_at_local FROM ? p WHERE p.id = photo_id) END WHERE ?",
						gorm.Expr(photosTable), updateWhere).Error)
			
			default:
				//Keeping taken_at_local, original request 
				Log("files", "regenerate photo_taken_at",
					Db().Exec("UPDATE files SET photo_taken_at = (SELECT p.taken_at_local FROM ? p WHERE p.id = photo_id) WHERE ?",
						gorm.Expr(photosTable), updateWhere).Error)	
		}
```

(sorry in advance for ugly code, these are just rough drafts to illustrate the idea)
